File: lib/py/uapi/internal/binary/ClientBinaryEncode.py
from typing import List, Dict, Optional
from uapi.ClientBinaryStrategy import ClientBinaryStrategy
from uapi.internal.binary.BinaryEncoding import BinaryEncoding
from uapi.internal.binary.BinaryEncoderUnavailableError import BinaryEncoderUnavailableError
from uapi.internal.binary.EncodeBody import encode_body
from uapi.internal.binary.PackBody import pack_body


def client_binary_encode(message: List[object], recent_binary_encoders: Dict[int, 'BinaryEncoding'],
                         binary_checksum_strategy: 'ClientBinaryStrategy') -> List[object]:
    headers = message[0]
    message_body = message[1]
    force_send_json = headers.pop("_forceSendJson", None)

    headers["bin_"] = binary_checksum_strategy.get_current_checksums()

    if force_send_json == True:
        raise BinaryEncoderUnavailableError()

    if len(recent_binary_encoders) > 1:
        raise BinaryEncoderUnavailableError()

    binary_encoder_optional = next(iter(recent_binary_encoders.values()), None)
    if not binary_encoder_optional:
        raise BinaryEncoderUnavailableError()
    binary_encoder = binary_encoder_optional

    encoded_message_body = encode_body(message_body, binary_encoder)

    final_encoded_message_body = pack_body(encoded_message_body) if headers.get(
        "_pac") == True else encoded_message_body

    return [headers, final_encoded_message_body]

File: lib/ts/uapi/internal/binary/ClientBinaryEncode.ts
import { BinaryEncoding } from 'uapi/internal/binary/BinaryEncoding';
import { BinaryEncoderUnavailableError } from 'uapi/internal/binary/BinaryEncoderUnavailableError';
import { ClientBinaryStrategy } from 'uapi/ClientBinaryStrategy';
import { encodeBody } from 'uapi/internal/binary/EncodeBody';
import { packBody } from 'uapi/internal/binary/PackBody';

export function clientBinaryEncode(
    message: any[],
    recentBinaryEncoders: Map<number, BinaryEncoding>,
    binaryChecksumStrategy: ClientBinaryStrategy,
): any[] {
    const headers = message[0] as Record<string, any>;
    const messageBody = message[1] as Record<string, any>;
    const forceSendJson = headers['_forceSendJson'];

    headers['bin_'] = binaryChecksumStrategy.getCurrentChecksums();

    if (forceSendJson === true) {
        throw new BinaryEncoderUnavailableError();
    }

    if (recentBinaryEncoders.size > 1) {
        throw new BinaryEncoderUnavailableError();
    }

    const binaryEncoder = [...recentBinaryEncoders.values()][0];

    if (!binaryEncoder) {
        throw new BinaryEncoderUnavailableError();
    }

    const encodedMessageBody = encodeBody(messageBody, binaryEncoder);

    let finalEncodedMessageBody: Map<any, any>;
    if (headers['_pac'] === true) {
        finalEncodedMessageBody = packBody(encodedMessageBody);
    } else {
        finalEncodedMessageBody = encodedMessageBody;
    }

    return [headers, finalEncodedMessageBody];
}
